### PR TITLE
MTKA-1488: blueprint local file upload

### PR DESCRIPTION
**Pull request by matthewguywright**
## Description

What changed and why?
Updated get remote blueprint to check for the path and use local or remote function accordingly.

Link to the JIRA ticket if available.
[https://wpengine.atlassian.net/browse/MTKA-1448](https://wpengine.atlassian.net/browse/MTKA-1448)

## Testing
Added tests for local file ZIP type and for empty contents.

## Documentation Changes
Updated verbiage for blueprint cli.

### CODE REVIEW

File: tests/integration/blueprints/test-blueprint-downloads.php [hunk 1]
```diff
@@ -1,9 +1,31 @@
...
+	 * @covers ::WPE\AtlasContentModeler\Blueprint\Fetch\get_local_blueprint
+	 */
+	public function test_get_local_blueprint_returns_valid_zip_file(): void {
+		if ( ! defined( 'FS_METHOD' ) ) {
+			define( 'FS_METHOD', 'direct' ); // Allows direct filesystem copy operations without FTP/SSH passwords. This only takes effect during testing.
+		}
+
+		$blueprint = get_local_blueprint( __DIR__ . '/test-data/acm-rabbits.zip' );
+		$this->assertNotWPError( $blueprint );
+		self::assertSame( 'application/zip', mime_content_type( $blueprint ) );
```
Comment (nickcernis) [inline tests/integration/blueprints/test-blueprint-downloads.php:26]: This is the cause of the test failure.

We could try getting the mime type from the file contents in one of the ways suggested here: https://stackoverflow.com/questions/4671672/how-to-get-the-mime-type-of-a-file-after-using-file-get-contents-from-a-remote-s

Comment (nickcernis) [reply tests/integration/blueprints/test-blueprint-downloads.php:26]: @matthewguywright This is working locally for me:

```php
public function test_get_local_blueprint_returns_valid_zip_file(): void {
		if ( ! defined( 'FS_METHOD' ) ) {
			define( 'FS_METHOD', 'direct' ); // Allows direct filesystem copy operations without FTP/SSH passwords. This only takes effect during testing.
		}

		$blueprint = get_local_blueprint( __DIR__ . '/test-data/acm-rabbits.zip' );
		$this->assertNotWPError( $blueprint );

		$file_info = new finfo( FILEINFO_MIME );
		$mime_type = $file_info->buffer( $blueprint );
		self::assertSame( 'application/zip; charset=binary',  $mime_type );
	}
```

**Comment by nickcernis** [comment]
@matthewguywright Please could you add an entry to the changelog?

Apart from that and the test correction above this is looking good!

I found a couple of small docs things that seem worth updating too:

1. The WP-CLI synopsis and url here need the new synopsis and option (can copy the inline docs from the code where we renamed url to path):

https://github.com/wpengine/atlas-content-modeler/blob/c40dfcaa68734d366a9136f0ce80a42e573c78a6/docs/wp-cli/index.md?plain=1#L19-L28

2. The Blueprints doc here could be clarified to make it clear a local file path now works:

https://github.com/wpengine/atlas-content-modeler/blob/c40dfcaa68734d366a9136f0ce80a42e573c78a6/docs/blueprints/index.md?plain=1#L8

Perhaps:

> `wp acm blueprint import [url or path to zip file]` will import a blueprint from a publicly-accessible URL or local file.

…or whatever you think sounds best.

File: includes/blueprints/fetch.php [hunk 2]
```diff
@@ -10,7 +10,46 @@
...
+ *
+ * @param string $path The path to the file.
+ * @return string
+ */
+function get_blueprint( string $path ) {
+	if ( filter_var( $path, FILTER_VALIDATE_URL ) ) {
+		return get_remote_blueprint( $path );
+	} else {
+		return get_local_blueprint( $path );
+	}
```
Comment (mindctrl) [inline includes/blueprints/fetch.php:23]: Just a minor suggestion (from my editor) to reduce the nesting complexity a little.

```suggestion
	if ( filter_var( $path, FILTER_VALIDATE_URL ) ) {
		return get_remote_blueprint( $path );
	}

	return get_local_blueprint( $path );
```

File: includes/blueprints/fetch.php [hunk 3]
```diff
@@ -10,7 +10,46 @@
...
+			'acm_blueprint_invalid_file_path',
+			esc_html__( 'File path was invalid.', 'atlas-content-modeler' ),
+			[ 'status' => 400 ]
+		);
+	}
+
+	if ( ! is_readable( $path ) ) {
+		return new WP_Error(
+			'acm_blueprint_file_not_readable',
+			esc_html__( 'Received empty response body.', 'atlas-content-modeler' )
```
Comment (mindctrl) [inline includes/blueprints/fetch.php:44]: This was leftover from the copy paste I think. Probably should read something like this. Anyone have a better suggestion for error text?

```suggestion
			esc_html__( 'File not found or not readable.', 'atlas-content-modeler' )
```

**Review by mindctrl — CHANGES_REQUESTED** [review]

**Comment by nickcernis** [comment]
@matthewguywright Please could you add an entry to the changelog?

**Review by rfmeier — APPROVED** [review]
Nice work! 👍

**Comment by anthonyburchell** [comment]
This looks good and works great in my testing. I cant approve this PR but if I could I would 👍 lol

**Review by anthonyburchell — APPROVED** [review]
nvmd. I have the option. lol

**Review by nickcernis — APPROVED** [review]

**Review by mindctrl — APPROVED** [review]
Not sure why this still says I requested changes on this PR. Approving again and merging.

**Pull request merged**